### PR TITLE
[BEAM-11104] Allow self-checkpointing SDFs to return without finishing their restriction

### PR DESCRIPTION
**Pull request by jrmccluskey**
Fixes issue where a self-checkpointing SDF would return an error because its restriction was not fully processed in the execution. This prevents self-checkpointing SDFs from returning properly. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by lostluck — APPROVED** [review]
We should follow this up with a PR with unit tests. Coverage is only half of a testing story, expecially when complex if-conditions are at odds.

**Comment by lostluck** [comment]
Or more likely, include them in this PR, since this breaks the GoPortable precommit.

**Comment by jrmccluskey** [comment]
Yeah I see what happened on the break, val is not guaranteed to be non-nil. Let me work on the structure here...

**Comment by jrmccluskey** [comment]
The pubsubx tests are timing out when run for codecov purposes but are passing on the same branch on my local machine

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go [hunk 1]
```diff
@@ -174,17 +174,22 @@ func (n *ParDo) processSingleWindow(mainIn *MainInput) (sdf.ProcessContinuation,
...
 		// We do not forward a ProcessContinuation on its own
 		if val.Elm == nil {
 			return val.Continuation, nil
 		}
 		return val.Continuation, n.Out[0].ProcessElement(n.ctx, val)
+	} else {
+		if mainIn.RTracker != nil && !mainIn.RTracker.IsDone() {
+			return nil, rtErrHelper(mainIn.RTracker.GetError())
+		}
 	}
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/pardo.go:193]: Style nit: If the previous clause ends with a return, drop the else.

**Review by lostluck — CHANGES_REQUESTED** [review]
Since we're no longer in a hurry, please add a unit test.  (Unless the branch is in a state making that very difficult, at which point I can wait for the next PR.)

**Comment by jrmccluskey** [comment]
Codecov is being extremely silly with this one. I assure you that adding a few unit tests did not add nearly 5000 lines of untested code

**Comment by damccorm** [comment]
> Codecov is being extremely silly with this one. I assure you that adding a few unit tests did not add nearly 5000 lines of untested code

This is one of the weirder ones - its blaming the drop on registration.go - since that should be excluded from codecov by my PR, I was curious and dug in and registration.go doesn't even exist at your last commit... - https://github.com/apache/beam/tree/b6b7d8a837e42d8e8ce37b7940ec422d4c4c9b9f/sdks/go/pkg/beam/registration - apparently you have 4% coverage though ¯\_(ツ)_/¯

I'm reasonably confident that merging in the latest changes from master would fix the issue though

**Pull request merged**